host/sgxload.c

```diff
@@ -767,41 +732,48 @@ oe_result_t oe_sgx_delete_enclave(oe_enclave_t* enclave)
     if (!enclave)
         OE_RAISE(OE_INVALID_PARAMETER);
 
+    /* free allocate memory. */
+    result = _sgx_free_enclave_memory(
+        (void*)enclave->addr, enclave->size, enclave->simulate);
+
+done:
+
+    return result;
+}
+
+oe_result_t _sgx_free_enclave_memory(
+    void* addr,
+    size_t size,
+    bool is_simulation)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
 #if defined(__linux__)
 
-/* FLC Linux needs to call `enclave_delete` in SGX mode. */
 #if defined(OE_USE_LIBSGX)
-    if (!enclave->simulate)
+
+    if (!is_simulation)
     {
         uint32_t enclave_error = 0;
-        if (!enclave_delete((void*)enclave->addr, &enclave_error))
+        if (!enclave_delete((void*)addr, &enclave_error))
             OE_RAISE(OE_PLATFORM_ERROR);
         if (enclave_error != 0)
             OE_RAISE(OE_PLATFORM_ERROR);
+
+        result = OE_OK;
+    done:
+        return result;
     }
     else /* FLC simulation mode needs to munmap. */
 #endif
     {
-        /* Non-FLC Linux and simulation mode both allocate memory. */
-        munmap((void*)enclave->addr, enclave->size);
+        munmap(addr, size);
     }
 
 #elif defined(_WIN32)
-    /* SGX enclaves can be freed with `VirtualFree` with the `MEM_RELEASE`
-     * flag. We can't do this for simulation mode because `MEM_RELEASE`
-     * requires `enclave->addr` to be the same one returned by `VirtualAlloc`,
-     * which is often not the case due to the enclave address alignment
-     * requirements. We use `MEM_DECOMMIT` instead. */
-    if (!enclave->simulate)
-        VirtualFree((void*)enclave->addr, 0, MEM_RELEASE);
-    else
-        VirtualFree((void*)enclave->addr, enclave->size, MEM_DECOMMIT);
-
+    VirtualFree((void*)addr, 0, MEM_RELEASE);
 #endif
 
     result = OE_OK;
-
-done:
```

Comment:
Why was this moved? We would like to explicitly avoid multiple exits in the routine.

```c
oe_result_t _sgx_free_enclave_memory(
    void* addr,
    size_t size,
    bool is_simulation)
{
    oe_result_t result = OE_UNEXPECTED;

#if defined(__linux__)

#if defined(OE_USE_LIBSGX)
    if (!is_simulation)
    {
        uint32_t enclave_error = 0;
        if (!enclave_delete(addr, &enclave_error))
            OE_RAISE(OE_PLATFORM_ERROR);
        if (enclave_error != 0)
            OE_RAISE(OE_PLATFORM_ERROR);
    }
    else /* FLC simulation mode needs to munmap. */
#endif
    {
        munmap(addr, size);
    }

#elif defined(_WIN32)
    VirtualFree(addr, 0, MEM_RELEASE);
#endif

    result = OE_OK;

done:
    return result;
}
```

Reply:
If that's the case, then I would simplify the return structure to avoid using error codes directly:
```c
oe_result_t _sgx_free_enclave_memory(
    void* addr,
    size_t size,
    bool is_simulation)
{
#if defined(__linux__)

#if defined(OE_USE_LIBSGX)
    if (!is_simulation)
    {
        uint32_t enclave_error = 0;
        if (!enclave_delete(addr, &enclave_error) ||
            enclave_error != 0)
            return OE_PLATFORM_ERROR;
    }
    else /* FLC simulation mode needs to munmap. */
#endif
    {
        munmap(addr, size);
    }

#elif defined(_WIN32)
    VirtualFree(addr, 0, MEM_RELEASE);
#endif

    return OE_OK;
}
```